pkg/tools/benchmark/job/service/insert.go

```diff
@@ -19,58 +19,73 @@ package service
 
 import (
 	"context"
+	"math"
 	"strconv"
 
 	"github.com/vdaas/vald/apis/grpc/v1/payload"
+	"github.com/vdaas/vald/internal/errgroup"
 	"github.com/vdaas/vald/internal/errors"
 	"github.com/vdaas/vald/internal/log"
-	"github.com/vdaas/vald/internal/net/grpc/status"
 )
 
 func (j *job) insert(ctx context.Context, ech chan error) error {
 	log.Info("[benchmark job] Start benchmarking insert")
 	// create data
-	vecs := j.genVec(j.dataset)
+	vecs := j.hdf5.GetByGroupName(j.dataset.Group)
 	cfg := &payload.Insert_Config{
 		SkipStrictExistCheck: j.insertConfig.SkipStrictExistCheck,
 	}
 	if j.timestamp > int64(0) {
 		cfg.Timestamp = j.timestamp
 	}
-	for i := 0; i < len(vecs); i++ {
-		log.Infof("[benchmark job] Start insert: iter = %d", i)
-		err := j.limiter.Wait(ctx)
-		if err != nil {
-			if errors.Is(err, context.Canceled) {
-				return errors.Join(err, context.Canceled)
-			}
-			ech <- err
-		}
-		res, err := j.client.Insert(ctx, &payload.Insert_Request{
-			Vector: &payload.Object_Vector{
-				Id:     strconv.Itoa(i),
-				Vector: vecs[i],
-			},
-			Config: cfg,
-		})
-		if err != nil {
-			select {
-			case <-ctx.Done():
+	eg, egctx := errgroup.New(ctx)
+	eg.Limitation(j.concurrencyLimit)
+	for i := j.dataset.Range.Start; i <= j.dataset.Range.End; i++ {
+		iter := i
+		eg.Go(func() error {
+			log.Debugf("[benchmark job] Start insert: iter = %d", iter)
+			err := j.limiter.Wait(egctx)
```

Comment:
I am not sure, but is it better the `j.limiter.Wait` function is called before the goroutine is generated? I think most of the goroutines will be in a waiting state.
```suggestion
	for i := j.dataset.Range.Start; i <= j.dataset.Range.End; i++ {
		iter := i
		err := j.limiter.Wait(egctx)
		if err != nil {
		
		}
		eg.Go(func() error {
			log.Debugf("[benchmark job] Start insert: iter = %d", iter)
```